gnomad/resources/resource_utils.py

```diff
@@ -373,9 +373,47 @@ def __init__(self, default_version: str, versions: Dict[str, BlockMatrixResource
         super().__init__(default_version, versions)
 
 
+class ResourceNotAvailable(Exception):
+    """Exception raised if a resource is not available from the selected source."""
+
+
 class GnomadPublicResource(BaseResource, ABC):
     """Base class for the gnomAD project's public resources."""
 
+    def __init_subclass__(cls, *, read_resource_methods: Iterable[str] = []) -> None:
+        super().__init_subclass__()
+
+        # Some resources may not be available from all sources due to delays in syncing, etc.
+        # This wraps all methods that read the resource and adds a check for if the resource
+        # is available from the selected source. If the resource is not available, this
+        # throws a more helpful error than if the read were simply allowed to fail.
+        def _wrap_read_resource_method(method_name):
+            original_method = getattr(cls, method_name)
+
+            @wraps(original_method)
+            def read_resource(self, *args, **kwargs):
+                # If one of the known sources is selected, check if the resource is available.
+                # For custom sources, skip the check and attempt to read the resource.
+                resource_source = gnomad_public_resource_configuration.source
+                if (
+                    isinstance(resource_source, GnomadPublicResourceSource)
+                    and resource_source != GnomadPublicResourceSource.GNOMAD
+                ):
+                    if not self.is_resource_available():
```

Comment:
Do you think it makes sense to check if the gnomAD resource exists here( given how occasionally the repo lags behind buckets)?  I'm wondering if it makes sense to add functionality to load from the gnomAD source on failure? Or if that gets more complicated as we add other cloud providers public buckets.

Reply:
This was assuming that resources will always be present in the gnomAD bucket, since they are copied from there to the cloud provider buckets.

I did not consider the case where an older version of this code points to a resource that no longer exists in the gnomAD buckets. Ideally, those URLs would not change but, practically, they do. That's a good point. We could catch that and suggest users update to the most recent version of gnomad_methods.

I'll claim that we should not automatically load from the gnomAD source if the resource doesn't exist. Most Hail resources are in the requester pays bucket, so if a user has configured their pipeline to use one of the cloud provider buckets, automatically falling back to the gnomAD buckets could result in unexpected egress/operation costs for that user. Also, they may not have Hail configured to allow requester pays access. I think the best we can do is point them towards the solution.

Reply:
I did not think of the cost associated with the requester-pays bucket, it absolutely makes sense to not automatically load the gnomAD source.

Reply:
Filed #390 for this.